apollo-router/src/plugins/telemetry/otlp.rs

```diff
@@ -0,0 +1,240 @@
+    Url(Url),
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub enum EndpointDefault {
+    Default,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Default, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct HttpExporter {
+    pub headers: Option<HashMap<String, String>>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, Default, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct GrpcExporter {
+    #[serde(flatten)]
+    pub tls_config: Option<TlsConfig>,
+    #[serde(
+        deserialize_with = "header_map_serde::deserialize",
+        serialize_with = "header_map_serde::serialize",
+        default
+    )]
+    #[schemars(schema_with = "option_metadata_map", default)]
+    pub metadata: Option<MetadataMap>,
+}
+
+fn option_metadata_map(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+    Option::<HashMap<String, Value>>::json_schema(gen)
+}
+
+#[derive(Debug, Clone, Default, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+pub struct TlsConfig {
+    domain_name: Option<String>,
+    ca: Option<Secret>,
+    cert: Option<Secret>,
+    key: Option<Secret>,
+}
+
+impl TryFrom<&TlsConfig> for tonic::transport::channel::ClientTlsConfig {
+    type Error = BoxError;
+
+    fn try_from(config: &TlsConfig) -> Result<ClientTlsConfig, Self::Error> {
+        ClientTlsConfig::new()
+            .with(&config.domain_name, |b, d| b.domain_name(d))
+            .try_with(&config.ca, |b, c| {
+                Ok(b.ca_certificate(tonic::transport::Certificate::from_pem(c.read()?)))
+            })?
+            .try_with(
+                &config.cert.clone().zip(config.key.clone()),
+                |b, (cert, key)| {
+                    Ok(b.identity(tonic::transport::Identity::from_pem(
+                        cert.read()?,
+                        key.read()?,
+                    )))
+                },
+            )
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub enum Secret {
+    Env(String),
+    File(PathBuf),
+}
+
+impl Secret {
+    pub fn read(&self) -> Result<String, ConfigurationError> {
+        match self {
+            Secret::Env(s) => std::env::var(s).map_err(ConfigurationError::CannotReadSecretFromEnv),
+            Secret::File(path) => {
+                std::fs::read_to_string(path).map_err(ConfigurationError::CannotReadSecretFromFile)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub enum Protocol {
+    Grpc,
+    Http,
+}
+
+impl Default for Protocol {
+    fn default() -> Self {
+        vec![Protocol::Grpc, Protocol::Http]
+            .get(0)
+            .expect("at least one feature must be present for otel, qed")
+            .clone()
+    }
+}
+
+mod header_map_serde {
```

Comment:
I think it should be `metadata_map_serde`

Reply:
Fixed